src/libprojectM/MilkdropPresetFactory/Parser.cpp

```diff
@@ -1355,63 +1363,93 @@ int Parser::parse_float(std::istream &  fs, float * float_ptr)
 }
 
 /* Parses a per frame equation. That is, interprets a stream of data as a per frame equation */
-PerFrameEqn * Parser::parse_per_frame_eqn(std::istream &  fs, int index, MilkdropPreset * preset)
+PerFrameEqn* Parser::parse_per_frame_eqn(std::istream& fs, int index, MilkdropPreset* preset)
 {
 
-  char string[MAX_TOKEN_SIZE];
-  Param * param;
-  PerFrameEqn * per_frame_eqn;
-  Expr * gen_expr;
+    Param* param{ nullptr };
+    Expr* gen_expr{ nullptr };
+    PerFrameEqn* per_frame_eqn{ nullptr };
 
+    std::array<char, MAX_TOKEN_SIZE> string{ 0 };
 
-  if (parseToken(fs, string) != tEq)
-  {
-    if (PARSE_DEBUG) printf("parse_per_frame_eqn: no equal sign after string \"%s\" (LINE %d)\n", string, line_count);
-    return NULL;
-  }
+    token_t token = parseToken(fs, string.data());
 
-  /* Find the parameter associated with the string, create one if necessary */
-  if ((param = ParamUtils::find(string, &preset->builtinParams, &preset->user_param_tree)) == NULL)
-  {
-    return NULL;
-  }
-  if (PARSE_DEBUG) std::cerr << "parse_per_frame_eqn: parameter \"" << param->name << "\" retrieved (LINE" << line_count << ")" << std::endl;
-  /* Make sure parameter is writable */
-  if (param->flags & P_FLAG_READONLY)
-  {
-    if (PARSE_DEBUG) std::cerr << "parse_per_frame_eqn: parameter \"" << param->name << "\" %s is marked as read only (LINE " << line_count << ")" << std::endl;
-    return NULL;
-  }
+    if (token == tEOL)
+    {
+        return nullptr;
+    }
 
-  /* Parse right side of equation as an expression */
-  if ((gen_expr = parse_gen_expr(fs, NULL, preset)) == NULL)
-  {
-    if (PARSE_DEBUG) printf("parse_per_frame_eqn: equation evaluated to null (LINE %d)\n", line_count);
-    return NULL;
-  }
+    if (token != tEq)
+    {
+        if (PARSE_DEBUG)
+        {
+            printf("parse_per_frame_eqn: no equal sign after string \"%s\" (LINE %d)\n", string.data(), line_count);
+        }
+        return nullptr;
+    }
+
+    /* Find the parameter associated with the string, create one if necessary */
+    if ((param = ParamUtils::find(string.data(), &preset->builtinParams, &preset->user_param_tree)) == nullptr)
+    {
+        return NULL;
+    }
+    if (PARSE_DEBUG)
+    {
+        std::cerr << "parse_per_frame_eqn: parameter \"" << param->name << "\" retrieved (LINE" << line_count << ")"
+                  << std::endl;
+    }
+    /* Make sure parameter is writable */
+    if (param->flags & P_FLAG_READONLY)
```

Comment:
This `if` is a bit different from the one below

Reply:
Hum... yeah, I did not try to understand this code at all. This function was just reformatted and I replaced a few `NULL`s with `nullptr`. This line was line 1380 previously, it just moved down a bit.

I think you can see here why it's just better to rewrite this file reader from scratch than trying to "fix" it. Reimplementation will first read all the lines into a map of key/value pairs, then use a key prefix to retrieve init values or code blocks, more or less the same as Milkdrop does it. The existing parser here is just way too complex.